### PR TITLE
refactoring jellyfish-address

### PR DESCRIPTION
**Pull request by ivan-zynesis**
#### What kind of PR is this?:

/kind feature

#### What this PR does / why we need it:
better code coverage
less logical branches

#### Which issue(s) does this PR fixes?:

Fixes #

### CODE REVIEW

File: packages/jellyfish-address/src/p2wsh.ts [hunk 1]
```diff
@@ -64,4 +45,38 @@ export class P2WSH extends Bech32Address {
...
+  /**
+   * @param {string} raw jellyfish p2wpkh (bech32 address) string
+   * @throws when decoded prefix is not found in DeFiChain ecosystem (mainnet / testnet / regtest)
+   * @returns {P2WSH}
+   */
+  static from (raw: string): P2WSH {
+    let valid: boolean
+    let prefix: string
+    let data: string = ''
+    try {
```
[the remark this answers is not in the document]
Comment (ivan-zynesis) [reply packages/jellyfish-address/src/p2wsh.ts:60]: multiple "similar blocks" found

p2wsh similar to p2wpkh (bech32), however, refactoring them to parent class with require abstract methods like `getDataLength()` / `getExpectedPrefixes()`

File: packages/jellyfish-address/src/p2sh.ts [hunk 2]
```diff
@@ -1,45 +1,61 @@
...
+    let buffer: Buffer | undefined
+    let valid = false
+    try {
+      const decoded = Bs58.toHash160(utf8String)
+      buffer = decoded.buffer
+      network = [MainNet, TestNet, RegTest].find(net => net.scriptHashPrefix === decoded.prefix)
+      valid = network !== undefined
+    } catch {
+      // non b58 string, invalid address
+    }
```
Comment (canonbrother) [inline packages/jellyfish-address/src/p2sh.ts:58]: `try.. catch` without handling??

File: packages/jellyfish-address/src/p2wpkh.ts [hunk 3]
```diff
@@ -1,71 +1,68 @@
...
+    try {
+      const decoded = bech32.decode(raw)
+      const trimmedVersion = decoded.words.slice(1)
+      data = Buffer.from(bech32.fromWords(trimmedVersion))
+
+      network = [MainNet, TestNet, RegTest].find(net => net.bech32.hrp === decoded.prefix)
+      valid = data.length === P2WPKH.PUB_KEY_HASH_LENGTH && network !== undefined
+    } catch (e) {
+      // invalid address, fail to decode bech32
+    }
```
Comment (canonbrother) [inline packages/jellyfish-address/src/p2wpkh.ts:64]: same with https://github.com/DeFiCh/jellyfish/pull/302/files#r642516402

File: packages/jellyfish-address/src/p2sh.ts (hunk 2 above)
Comment (ivan-zynesis) [reply packages/jellyfish-address/src/p2sh.ts:58]: Intended. 

```
const addressStr = takeUserInput()
const address = DeFiAddress.from(addressStr)
```
It try instantiate multiple type of address and return if any valid. Return one with `valid` flag false even none is possible.

**Comment by fuxingloh** [comment]
this is still too complicated in my opinion, let keep it open for now

**Comment by fuxingloh** [comment]
This PR will be superseded by the introduction of #670, I will refactor `jellyfish-address` when #670 is done to simplify address logic.

**Pull request closed without merging**